### PR TITLE
Use separate compile/link steps for bootstrap-ekam

### LINKED ISSUE #26: ekam-bootstrap can't be ccached
On my machine, compiling the bootstrap ekam takes 88sec. On a github runner, it's around 25sec. It could be sped up by ccache, but ccache can't compile multiple source files on one commandline. I'd like to replace the bootstrap's Makefile recipe to compile each source file separately and then link them together.

### PR DESCRIPTION
**Pull request by abliss**
This should enhance parallelism and cachability for this operation.

Fixes #26 .

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -40,15 +40,22 @@ deps/capnproto:
...
     grep -v KqueueEventManager | grep -v PollEventManager | \
     grep -v ProtoDashboard | grep -v ekam-client | grep -v _test)
 
-bin/ekam-bootstrap: $(SOURCES)
+OBJ_DIR := tmp
+OBJECTS=$(addprefix $(OBJ_DIR)/, $(SOURCES:.cpp=.o))
+
+$(OBJ_DIR)/%.o: src/%.cpp
+	@mkdir -p $(@D)
+	$(CXX) -Isrc -fPIC -std=c++14 -pthread -o $@ -c $<
```
Comment (zenhack) [inline Makefile]: Since it's used in more than one place now, let's factor cflags out into a variable.

**Comment by kentonv** [comment]
Hmm, in theory this makefile is now unsound because it doesn't recognize when header changes should force an object to be rebuilt.

Since the goal here seems to be to enable caching (specifically for Sandstorm), would it make sense to look into caching the final ekam binary itself?

Ideally the sandstorm build would cache ekam build outputs keyed by git revision. I've actually written a script to do this for Cloudflare Workers; I could maybe adapt a version of that for Sandstorm.

**Comment by abliss** [comment]
Thanks. When you say "is now unsound," do you mean that it's more unsound
after my change than before? If so, could you please spell out how exactly
how? (E.g. could you describe a scenario where the previous makefile would
produce a correct binary, but after my changes it would produce an
incorrect one?)

It's true that a primary motivator for this change was to speed up clean
(but ccached) sandstorm builds. However, for my part, I would personally
prefer to see fewer (rather than more) custom-made scripts for the build. I
much prefer to see an accurate (and granular) dependency recorded in a
standard format (e.g. Makefile) whenever possible.

**Comment by zenhack** [comment]
The failure mode I assume @kentonv is referring to is, if you change a header file that one of the ekam src files includes, a rebuild will not consider the corresponding object file to be out of date.

(the fact that I didn't spot that tells me I'm a bit rusty :/)

+1 for less magic if we can manage it.

to solve the current soundness issue, you could the compilers `-MD` flag, and add an include directive to the Makefile.

**Comment by kentonv** [comment]
I guess you could make the point that the existing makefile doesn't list headers in the dependencies of this rule, so has a similar problem (though not really the same -- it'll never produce a binary that crashes due to ABI inconsistencies between objects, it just might not rebuild when it should).

But I think going all the way and writing a Makefile that uses `-MD` to track dependencies is going too far. The goal of this bootstrap rule was to be as simple as possible.

I haven't used ccache before. How exactly will this work? Do we need to stand up a shared ccache server? Or does each developer maintain their own private ccache? How does this work in CI?

**Comment by abliss** [comment]
I added all headers as dependencies to all objects.

**Comment by kentonv** [comment]
I guess this ended up becoming official in the Sandstorm build so... better merge it.

**Pull request merged**